packages/transformer/src/DetachedExportElementAspectsStrategy.ts

```diff
@@ -0,0 +1,108 @@
+  public async exportAllElementAspects(): Promise<void> {
+    await this.exportAspectsLoop<ElementUniqueAspect>(ElementUniqueAspect.classFullName, async (uniqueAspect) => {
+      const isInsertChange = this.aspectChanges?.insertIds.has(uniqueAspect.id) ?? false;
+      const isUpdateChange = this.aspectChanges?.updateIds.has(uniqueAspect.id) ?? false;
+      const doExport = this.aspectChanges === undefined || isInsertChange || isUpdateChange;
+      if (doExport) {
+        const isKnownUpdate = this.aspectChanges ? isUpdateChange : undefined;
+        this.handler.onExportElementUniqueAspect(uniqueAspect, isKnownUpdate);
+        await this.handler.trackProgress();
+      }
+    });
+
+    let batchedElementMultiAspects: ElementMultiAspect[] = [];
+    await this.exportAspectsLoop<ElementMultiAspect>(ElementMultiAspect.classFullName, async (multiAspect) => {
+      if (batchedElementMultiAspects.length === 0) {
+        batchedElementMultiAspects.push(multiAspect);
+        return;
+      }
+
+      // element id changed so all element's aspects are in the array and can be exported
+      if (batchedElementMultiAspects[0].element.id !== multiAspect.element.id) {
+        this.handler.onExportElementMultiAspects(batchedElementMultiAspects);
+        await this.handler.trackProgress();
+        batchedElementMultiAspects = [];
+      }
+
+      batchedElementMultiAspects.push(multiAspect);
+    });
+
+    if (batchedElementMultiAspects.length > 0) {
+      // aspects that are left in the array have not been exported
+      this.handler.onExportElementMultiAspects(batchedElementMultiAspects);
+      await this.handler.trackProgress();
+    }
+  }
+
+  private async exportAspectsLoop<T extends ElementAspect>(baseAspectClass: string, exportAspect: (aspect: T) => Promise<void>) {
+    for await (const aspect of this.queryAspects<T>(baseAspectClass)) {
+      if (!this.shouldExportElementAspect(aspect)) {
+        continue;
+      }
+
+      await exportAspect(aspect);
+    }
+  }
+
+  private async *queryAspects<T extends ElementAspect>(baseElementAspectClassFullName: string) {
+    const aspectClassNameIdMap = new Map<string, Id64String>();
+
+    const getAspectClassesSql = `SELECT DISTINCT ECClassId as classId, (ec_classname(ECClassId)) as className FROM ${baseElementAspectClassFullName}`;
+    const aspectClassesQueryReader = this.sourceDb.createQueryReader(getAspectClassesSql);
+    const aspectClassesAsyncQueryReader = ensureECSqlReaderIsAsyncIterableIterator(aspectClassesQueryReader);
+    for await (const rowProxy of aspectClassesAsyncQueryReader) {
+      const row = rowProxy.toRow();
+      aspectClassNameIdMap.set(row.className, row.classId);
+    }
+
+    for (const [className, classId] of aspectClassNameIdMap) {
+      if(this._excludedElementAspectClassFullNames.has(className))
+        continue;
+
+      const getAspectPropsSql = `SELECT * FROM ${className} WHERE ECClassId = :classId ORDER BY Element.Id`;
```

Comment:
It might be, but the reason for this ordering is that because `IModelImporter` and from it `IModelTransformer` requires that all element's `ElementMultiAspects` with the same class would be processed in one bunch. Ordering aspects by `ECClassId, ECInstanceId` won't ensure that all element's aspects of the same class will be exported in one bunch.

Ideally I would like to refactor that logic as well, but do it as a separate PR since there are more bugs that are present in that logic which are not related to memory optimizations or how exporter exports aspects (for example, it will miss to delete aspects in `targetDb`, if during `processChanges` there are no longer any aspects of that class in the `sourceDb`)

Reply:
reasonable, thanks, closing